### PR TITLE
[블랙잭] 권순찬 2주차 미션 제출합니다.

### PR DESCRIPTION
**Pull request by Kwon770**
Test code 작성과 TDD에 익숙해지고 있긴한데, GameController와 Game 도메인들은 테스트 코드 작성이 애매하네요. 원래 그런 것이겠죠..?

### CODE REVIEW

File: src/main/java/blackjack/domain/Game.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        return users;
+    }
+
+    public Dealer getDealer() {
+        return dealer;
+    }
+
+    public List<UserInfoDto> getUserInfoDtos() {
+        return users.stream().map(user ->
+                user.toDto()
```
Comment (hotheadfactory) [inline src/main/java/blackjack/domain/Game.java]: 일반적으로 User가 DTO에 의존하는 것보다 DTO가 User에게 의존하게 하는 것이 낫습니다. 도메인보다는 DTO를 변경할 일이 더 많기 때문입니다.

File: src/main/java/blackjack/domain/Player.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    public String getName() {
+        return name;
+    }
+
+    public OwnedCards getOwnedCards() {
+        return ownedCards;
+    }
+
+    public boolean getIsBlackjack() {
```
Comment (hotheadfactory) [inline src/main/java/blackjack/domain/Player.java]: 보통 boolean을 리턴하는 메소드의 네이밍 컨벤션은 `is~~` 입니다.
`isBlackjack`이 더 어울릴 것 같습니다! 다른 boolean을 리턴하는 메소드들도 마찬가지입니다.

File: src/main/java/blackjack/domain/Game.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        boolean didWanted = InputView.getYesOrNo().equals("y");
+        if (!didWanted) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public void judgePlayerBurst() {
+        for (User user : users) {
```
Comment (hotheadfactory) [inline src/main/java/blackjack/domain/Game.java]: 이런 곳에서 for문 대신 stream을 사용하면 어떤 장단점이 있을까요?

File: src/main/java/blackjack/domain/CardPack.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+
+public class CardPack {
+    public static final int CARD_SHAPE_COUNT = 4;
+    public static final int CARD_VALUE_COUNT = 13;
+
+    List<Card> cards = new ArrayList<>();
+    int selectedIndex = -1;
+
+    public CardPack() {
```
Comment (hotheadfactory) [inline src/main/java/blackjack/domain/CardPack.java]: 만약에 어디선가 카드 팩 생성자에 또 접근해서 여러 개의 카드 팩이 만들어진다면 어떻게 해야 할까요?

**Review by hotheadfactory — CHANGES_REQUESTED** [review]
수고하셨습니다.
별로 많이 고쳐야 할 부분은 없지만 코멘트 달아드린 부분에 대해 고민해 봅시다!

File: src/main/java/blackjack/domain/Game.java (hunk 3 above)
Comment (Kwon770) [reply src/main/java/blackjack/domain/Game.java]: 피드백해주신 것을 바탕으로 forEach와 메소드 참조를 이용하여 수정해본 결과 코드의 줄수가 확연히 줄어들었지만, for문에 비해 한 눈에 확들어오지는 않았습니다. 이것이 의도하신 장단점이 맞을까요?

File: src/main/java/blackjack/domain/CardPack.java (hunk 4 above)
Comment (Kwon770) [reply src/main/java/blackjack/domain/CardPack.java]: CardPack을 생성과 관리는 Game.java가 하기때문에 무관한 곳에서 CardPack을 여러개 생성한다면 게임에 문제가 발생한 것 같지않고, 확장성을 위해 static화한다는 선택지는 옳지않다고 생각이 듭니다. 말씀해주신 문제 상황은 어떤 것인지 이해가지 않습니다... 😢

File: src/main/java/blackjack/domain/Player.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    public void judgeBlackjack() {
+        isBlackjack = ownedCards.isBlackjack();
+    }
+
+    public void judgeBurst() {
+        isBurst = ownedCards.isBurst();
+    }
+
+    public void earnBet(int bet) {
```
Comment (hotheadfactory) [inline src/main/java/blackjack/domain/Player.java:47]: 굳이 수익과 손해를 나누기보다, 손해 시 마이너스 금액을 더하는 식으로 구현하는 것이 복잡도를 줄이는 데 도움이 되지 않을까요?

File: src/test/java/CardTest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    void test_getTotalScore_with10and10and10() {
+        ownedCards.addCard(new Card(CardShape.CLOVER, CardGrade.JUMP));
+        ownedCards.addCard(new Card(CardShape.CLOVER, CardGrade.JUMP));
+        ownedCards.addCard(new Card(CardShape.CLOVER, CardGrade.JUMP));
+
+        Assertions.assertEquals(30, ownedCards.getScore());
+    }
+
+    @Test
+    @DisplayName("카드가 10/10/A 일때, 점수가 올바르게 계산되는지?")
```
Comment (hotheadfactory) [inline src/test/java/CardTest.java:59]: 바로 이렇게, 테스트를 할 때에는, 같은 메소드에 대해서도 여러 edge case를 테스트하는 것이 좋습니다!!

File: src/test/java/InputTest.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+class InputTest {
+
+    @Test
+    @DisplayName("이름을 담긴 정상적인 문자열이 리스트로 변환되는지?")
+    void test_parseNameString_withProperString() {
+        List<String> names = InputView.parseNameString("a,b,c");
```
Comment (hotheadfactory) [inline src/test/java/InputTest.java:14]: 이 부분은 원래 private method여야 하는 것을 오직 테스트만을 위해 public으로 변경시켜 테스트를 진행한 것으로 보이는데요,
과연 이렇게 하는 게 옳은지에 대해 고민해 보시는 걸 추천해 드릴게요, 정해진 답은 없지만, 고민해 볼만한 문제에요.

**Review by hotheadfactory — APPROVED** [review]
깔끔하게 잘 구현해 주셨네요! 수고하셨습니다!

피드백 드린 부분은 다음 미션에 녹여내 보세요!
